src/auxiliary/mpi_arrays.jl

```diff
@@ -0,0 +1,194 @@
+
+# TODO: MPI. We keep this module inside Trixi for now. When it stabilizes and
+#            turns out to be generally useful, we can consider moving it to a
+#            separate package with simple test suite and documentation.
+module TrixiMPIArrays
+
+using ArrayInterface: ArrayInterface
+using MPI: MPI
+
+using ..Trixi: Trixi, mpi_comm
+
+export TrixiMPIArray, local_length, local_copy
+
+
+# Dispatch etc.
+# The following functions have special dispatch behavior for `TrixiMPIArray`s.
+# - `wrap_array`:
+#   the wrapped array is wrapped again in a `TrixiMPIArray`
+# - `wrap_array_native`:
+#   should not be changed since it should return a plain `Array`
+# - `allocate_coefficients`:
+#   this handles the return type of initialization stuff when setting an IC
+#   with MPI
+# Besides these, we usually dispatch on MPI mesh types such as
+# `mesh::ParallelTreeMesh` or ``mesh::ParallelP4eestMesh`, since this is
+# consistent with other dispatches on the mesh type. However, we dispatch on
+# `u::TrixiMPIArray` whenever this allows simplifying some code, e.g., because
+# we can call a basic function on `parent(u)` and add some MPI stuff on top.
+"""
+    TrixiMPIArray{T, N} <: AbstractArray{T, N}
+
+A thin wrapper of arrays distributed via MPI used in Trixi.jl. The idea is that
+these arrays behave as much as possible as plain arrays would in an SPMD-style
+distributed MPI setting with exception of reductions, which are performed
+globally. This allows to use these arrays in ODE solvers such as the ones from
+OrdinaryDiffEq.jl, since vector space operations, broadcasting, and reductions
+are the only operations required for explicit time integration methods with
+fixed step sizes or adaptive step sizes based on CFL or error estimates.
+
+!!! warning "Experimental code"
+    This code is experimental and may be changed or removed in any future release.
+"""
+struct TrixiMPIArray{T, N, Parent<:AbstractArray{T, N}} <: AbstractArray{T, N}
+  u_local::Parent
+  mpi_comm::MPI.Comm
+  mpi_rank::Int
+  mpi_size::Int
+end
+
+function TrixiMPIArray(u_local::AbstractArray{T, N}) where {T, N}
+  # TODO: MPI. Hard-coded to MPI.COMM_WORLD for now
+  mpi_comm = MPI.COMM_WORLD
+  mpi_rank = MPI.Comm_rank(mpi_comm)
+  mpi_size = MPI.Comm_size(mpi_comm)
+  TrixiMPIArray{T, N, typeof(u_local)}(u_local, mpi_comm, mpi_rank, mpi_size)
+end
+
+
+# Custom interface and general Base interface not covered by other parts below
+Base.parent(u::TrixiMPIArray) = u.u_local
+Base.resize!(u::TrixiMPIArray, new_size) = resize!(parent(u), new_size)
+function Base.copy(u::TrixiMPIArray)
+  return TrixiMPIArray(copy(parent(u)), u.mpi_comm, u.mpi_rank, u.mpi_size)
+end
+
+Trixi.mpi_comm(u::TrixiMPIArray) = u.mpi_comm
+Trixi.mpi_rank(u::TrixiMPIArray) = u.mpi_rank
+Trixi.mpi_nranks(u::TrixiMPIArray) = u.mpi_size
```

Comment:
If we follow YAGNI, the latter two can also be removed. The first one can be removed if we undo the `mpi_comm(u)` stuff in the analysis and stepsize callbacks as suggested by me.

Reply:
Agreed.